src/coreclr/jit/lower.cpp

```diff
@@ -7029,6 +7050,76 @@ void Lowering::ContainCheckJTrue(GenTreeOp* node)
     cmp->gtFlags |= GTF_SET_FLAGS;
 }
 
+//------------------------------------------------------------------------
+// ContainCheckCompareChain : determine whether a chain of compares should be contained by the given parent.
+//
+// Work backwards through the chain until it is no longer valid or containable.
+//
+// Arguments:
+//    tree - the node. Either a compare or a tree of compares connected by ANDs.
+//    parent - the parent of the tree.
+//    earliest_valid - Returns the earliest valid compare node in the chain (if any).
+//
+// Return Value:
+//    True if the chain is valid
+//
+bool Lowering::ContainCheckCompareChain(GenTree* tree, GenTree* parent, GenTree** earliest_valid)
```

Comment:
I'm wondering if we can start using this in other places in this PR to test it without having to introduce the complicated if-conversion logic.
For example, should it not be beneficial to do this for any possible comparison/and node? E.g. code like
`bool b = (x < 5) & (y < 3)`;` might be able to use this even without if-conversion.

Reply:
In fact, thinking about it some more, this part of the PR seems orthogonal to `GT_SELECT`, there are several nodes that can be taught to consume flags without materializing a truth value in a register:
* `GT_AND` via `ccmp`
* `GT_SELECT` via `csel`
* `GT_JTRUE` via `cb`

Hopefully all of these will end up on the same plan (doesn't have to be in this PR, but would be nice if we could do something for `GT_AND` to at least test it out).

Reply:
> ensure we only do this when it is profitable

Agreed. Looking at the generated code, having a AND chain of three items results in the same number of instructions (and probably exactly the same performance). 

I set the code to assert when the AND chain was 4 or more and ran `superpmi replay` .... and nothing fired.Looking through some dumps, there are potentially some larger chains in the IR, but all of them have a CMP node that fails `IsSafeToContainMem()`, causing the chain to break.

(shorter chains are still useful if being consumed by a SELECT node)

So, it's probably not worth going down the AND chain route.


> but would be nice if we could do something for GT_AND to at least test it out

This PR is only the lowering stage downwards as I was trying to split the work into smaller chunks in order to upstream more easily. Would you prefer it if I put the higher level passes into this PR too? That way at least everything is getting tested.

Reply:
> Agreed. Looking at the generated code, having a AND chain of three items results in the same number of instructions (and probably exactly the same performance).

What are we generating? I would hope it would be fewer instructions already at 2 comparisons for simple cases.
For example: https://godbolt.org/z/jzaEj9587
With your change I was hoping RyuJIT would generate similar code to Clang here (ignoring the frame setup, which is required in .NET).

> Would you prefer it if I put the higher level passes into this PR too? That way at least everything is getting tested.

If you can push your `GT_AND` changes, even with `n=2`, it would be great -- we can trigger stress jobs and check that everything looks right. Make sure to add the optimization check as I described as it will allow us to run fuzzers on the change.

Reply:
That's true,  cmp+cset+cmp+cset+and should now become cmp+ccmp+cset.  Looks like these aren't triggering when they should. I'll have to debug a little more.

The longer chains that were worse was because I was allowing scenarios where the first op of the AND can be outside the chain. Before my code:
```
IN0046: 000128                    cmp     x23, x22
IN0047: 00012C                    cset    x0, eq
IN0048: 000130                    ldr     x1, [x20,#32]
IN0049: 000134                    cmp     x2, x1
IN004a: 000138                    cset    x1, eq
IN004b: 00013C                    and     w0, w0, w1
```
Here the first cmp ends up failing to join the chain (due to the `IsSafeToContainMem` failing).
The second cmp and and join into a chain.
This gives:
```
IN0046: 000128                    cmp     x23, x22
IN0047: 00012C                    cset    x0, eq.   (This failed the IsSafeToContainMem check)
IN0048: 000130                    ldr     x1, [x20,#32]
IN0049: 000134                    cmp     w0, #0      (get the value back into flags)
IN004a: 000138                    ccmp    x2, x1, 0, ne
IN004b: 00013C                    cset    x0, eq
```
That's clearly no better. But a longer chain with the same characteristics would have been fine. Sadly we don't find any.

Reply:
```
N014 ( 24, 20) [000010] J------N---                         *  EQ        int    $184
N012 ( 22, 17) [000008] -----------                         +--*  AND       int    $183
N008 ( 15, 12) [000002] -c---------                         |  +--*  EQ        int    $181
N005 (  3,  2) [000048] -----------                         |  |  +--*  LCL_VAR   int    V05 cse0         u:1 $180
N007 (  1,  2) [000001] -c---------                         |  |  \--*  CNS_INT   int    0 $40
N011 (  6,  4) [000005] Nc-------U-                         |  \--*  LE        int    $182
N009 (  1,  1) [000003] -----------                         |     +--*  LCL_VAR   int    V01 arg1         u:1 $c0
N010 (  1,  2) [000004] -c---------                         |     \--*  CNS_INT   int    9 $45
N013 (  1,  2) [000009] -----------                         \--*  CNS_INT   int    0 $40
```

Reply:
Ok, so the smaller chains were getting cleared by the various `JCC` and `TEST_EQ` optimisations in the lowering stage. I've stopped it from clearing the contained state, but there are issues....

Consider:
```
[000011] -----+-----                         *  JTRUE     void  
[000010] J------N---                         *  EQ        int    $184
[000008] -----------                         +--*  AND       int    $183
[000002] -c---------                         |  +--*  GT        int    $181
[000048] -----------                         |  |  +--*  LCL_VAR   int    V05 cse0         u:1 $180
[000001] -c---------                         |  |  \--*  CNS_INT   int    0 $40
[000005] Nc-------U-                         |  \--*  LE        int    $182
[000003] -----------                         |     +--*  LCL_VAR   int    V01 arg1         u:1 $c0
[000009] -----------                         \--*  CNS_INT   int    0 $40
```
The `GT` and `LE` have been lowered by the `AND` into a contained chain (shown via the 'c' flag).

Next the const compare lowering happens, causing the `EQ` to be replaced with a `TEST_EQ` and the `AND` and `CNS_INT` (000009) are removed.

Ending up with lowered nodes:
```
LE ...
GT ...
TEST_EQ
JTRUE
```

Without the anything set as contained, we'd generate:
```
cmp ...
cset x0, le
cmp ...
cset x1, gt
tst x0, x1
beq
```

Ideally, the code we want is:
```
cmp     ...
ccmp    ...., le
jgt
```
Note the jump needs to pull the condition code from the GT node (its grandchild).

The lowered IR above is probably not ideal for contained nodes. But what would be the correct version?

Best I can think of is:
```
LE ...
GT ...
AND
JTRUE
```
But this still requires the `JTRUE` to pull its condition code from the `GT`. Plus the `AND` needs to not generate into a register.
Ideally everything would be contained within the `JTRUE` - but that I suspect is quite a mess to rework.

On top of this, all the above goes away when I add in the `SELECT` support as that `JTRUE` node would be replaced by a `SELECT`. Alternative option would be to push the entire `SELECT` work without any `CCMP`, then go back to the `CCMP`s

Reply:
It sounds like there is a decision to be made here -- what is the most efficient way to generate this pattern of nodes. It does not matter how we represent the sequence of nodes, we still have to make that decision. The question is where and how we make it:
* By disallowing the `TEST_EQ` transformation if it undoes the conditional compare transformation done via containment (easiest way with current PR). For example, you can add a helper like `GenTree::IsConditionalCompare` that recognizes `GT_AND` with contained compares
* By replacing some nodes with potential conditional compare nodes that then breaks the recognition of the `TEST_EQ` pattern (requires reintroducing conditional compare nodes)

If this was the only question I would have a preference for the former, but I am not sure if your other concerns also apply to the contained `GT_AND` form.
It is also ok with me if you want to scope back the changes to only support the conditional compare chains as operands of `GT_SELECT` again. My main concern is that we are not going to have good test coverage, I am not sure how much coverage `GT_SELECT` alone will get.

> Note the jump needs to pull the condition code from the GT node (its grandchild).

Right, but is it not already the case for `GT_SELECT`? It gets the condition from `genCodeForContainedCompareChain`, which looks simple enough. Or am I misunderstanding?

> Ideally everything would be contained within the JTRUE - but that I suspect is quite a mess to rework.

I think long time this is what we wand to end up with -- `GT_JTRUE` able to consume conditions from the flags register in the same way that `GT_AND` and `GT_SELECT` would be able to in this change.
However, I agree we should leave it for a future PR and just take the not-as-efficient codegen for now (I assume there is no regression here, right?)

> On top of this, all the above goes away when I add in the SELECT support as that JTRUE node would be replaced by a SELECT

I don't see how `GT_SELECT` helps in the general case. I can only see that it would help if the successor blocks to the block containing `GT_JTRUE` are simple enough that the entire thing can be replaced with `GT_SELECT`. Can you elaborate what you mean?

Reply:
> It sounds like there is a decision to be made here -- what is the most efficient way to generate this pattern of nodes. It does not matter how we represent the sequence of nodes, we still have to make that decision.

Just to clarify, because I'm not sure my point was very clear. Right now the `TEST_EQ` transformation is assuming ahead of time that it is always the most optimal way to do it. If we instead had conditional compare nodes, we would be doing the opposite -- assuming ahead of time that the conditional compares are always the most optimal way. In reality we need to consider these transforms that conflict on a case by case basis and make a decision for each on what the best way is. So I think it is a good thing that we hit a situation like this.

Reply:
I also think it's best to disable the TEST_EQ optimisation if it sees the contained chain. Disabling that cause the JTRUE to JCC optimisation to kick in instead. Disabling that too, means we end up with:

```
cmp     ...
ccmp    ...., le
cset x0, gt
cbz x0, label
```

Which is much better than the code without my patch.

> I don't see how GT_SELECT helps in the general case. I can only see that it would help if the successor blocks to the block containing GT_JTRUE are simple enough that the entire thing can be replaced with GT_SELECT. Can you elaborate what you mean?

In most cases in the code above, the last two instructions will become a `csel`.
But, yes, there will be scenarios where we can't use `GT_SELECT`. In most of those cases though, we'll also fail to generate a compare chain. If it turns out there are more instances than I expect, then we can look at using the JCC in a follow on patch to replace the last two instructions in the code above with a `jgt`.

Reply:
> Which is much better than the code without my patch.

Indeed, looks great. I look forward to the PR that puts `JTRUE` on the same plan as this.
One thing I would like to see is a micro benchmark to make sure our expectation that this is better is correct.
The easy way to do that is via benchmark.NET where you can specify the old and new corerun with the `--corerun <old corerun> <new corerun>`.

> In most cases in the code above, the last two instructions will become a csel.

I would not expect that to be the case. Only for very limited forms of IR will we be able to generate `GT_SELECT` in this case, since it requires both successor blocks to be single assignments to the same variable without any other side effects. This is probably a common pattern, but not the the majority case pattern?

Reply:
FWIW, I'm fine with this getting low hits for now. I expect in .NET 8 we can enable it for JTRUE and expand boolean optimizations (I believe we do not transform `(a relop b) && (c relop d)` => `(a relop b) & (c relop d)` today), which should make the optimization much more impactful.